### ISSUE TITLE
mosh does not work with IPv6 when IPv4 is also available

### ISSUE THREAD

**Issue by nielsk** [blinksh/blink#269]
I have a host with an A- and an AAAA-record. ssh via IPv4 is closed to the outside world but IPv6 is open. 
When I use ssh with the Hostnamen to connect via blink, no problem. When I use the IPv6-address and mosh, no problem. 
When I use mosh and the host name, Blink tells me that it cannot find the remote host. 
And there is no "mosh -6" and I cannot declare that Blink should prefer IPv6 over IPv4. Please fix.

**Comment by nielsk** [comment]
When I enter in the config the IPv6-address for the host instead of the name, the Blink tells me "$IPv6-address: command not found"

This the only way to connect is via "mosh $IPv6-address". 

Please fix; this is my main jump-host.

**Comment by nielsk** [comment]
When I have an IPv6-address and try to a dual-stack host blink-shell cannot connect; same error as above. Apparently blink-shell even doesn't try the IPv4-address.

The only way I can connect now to my main hosts is by entering the IPv6-address by hand because my phone gets from my ISP automatically an IPv6-address

**Comment by carloscabanero** [comment]
Thanks for the report! Will try to take a look at it today.

**Comment by Eupeodes** [comment]
I just had the same problem on my server running debian jessie and mosh 1.24a, after upgrading mosh to 1.25 from the jessie-backports repo I can connect using ipv4 and ipv6 without any problems.

**Comment by nielsk** [comment]
I run mosh 1.3.0 on FreeBSD 10.3 and 11 while experiencing this problem.

**Comment by carloscabanero** [comment]
Still tracking this one, wanna fix it before pushing a new version with other bugs. One question, when you say "$IPv6-address: command not found" that basically means that we are getting connected but we are sending a command to the remote. How did you configure that on settings?

**Comment by nielsk** [comment]
I either entered the host name or the IPv6-address of the host. I did not configure any commands.

**Comment by nielsk** [comment]
![](http://i.imgur.com/OVpNOSU.jpg)
This is how I configured the host. And I use:
"mosh nessie" to connect to it

**Comment by carloscabanero** [comment]
There is an error there - which might be a UI problem too. The Mosh Server field allows to set the path to the server command (on OS X /opt/local/bin, linux might /usr/local/bin), useful for configuring that. So Blink is connecting to your server, but you are sending it your ipv6 as a command and that's why it is returning a "Command not found".

Just remove that, and put your IPv6 if necessary on the SSH Hostname.

**Comment by nielsk** [comment]
Bäm, it worked. Thanks. Because it said "server" I thought I had to put in there the address.

**Comment by Eupeodes** [comment]
For the field I think you can change the name to Server-path (perhaps with placeholder ‘optional’) (And maybe I should file this as a separate issue).

**Comment by carloscabanero** [comment]
Yeah, that is a great name suggestion, will do that. My head was all "Server Command Path".... and too long :D 

(Don't worry about the separate issue, I will keep this one open and fix it here.)

**Comment by Gunni** [comment]
I had this today, it turns out that this is documented behavior:

```
       --family=prefer-inet
              Similar to --family=all, but attempt connects to the IPv4 addresses first.  This is the default.
```

It's obviously backwards, it should happy eyeballs the connection, or prefer inet6.

For now, i suggest that `--family=all` be made the default.

**Comment by carloscabanero** [comment]
This is now implemented in our new Mosh client (Blink 17.2.). The default is using the resolution coming from LibSSH, which internally is using iOS methods that use Happy Eyeballs (https://developer.apple.com/forums/thread/47989). This should cover most of the situations. Then you can enforce with flag either IPv6 or IPv4 when necessary. Note this is on top of where the resolution happens (the remoteIP flag).
